### ISSUE TITLE
How to train ner_crf to extract 'title', 'description' like entities from a question ?

### ISSUE THREAD

**Issue by madhusudhannsn** [RasaHQ/rasa_nlu#797]
I am using RASA for my personal assistant which can help with google tasks, gmail etc., As part of this, I came across extracting task title and description from the question. For this, I trained a model using following examples with 'test title' as title entity.

> Please help me in creating a task with title test title
>  Help me in creating a task with title test title
>  Please create a task titled test title
>  Create a task with title test title
>  I would like to change the due date of task with title test title
>  Change the due date of a task titled test title
>  I would like to change the due date of task titled test title
>  Please delete the task with title test title
>  Please remove the task titled test title
>  I would like to delete task with title test title
>  Could you please delete the task titled test title

I expected ner_crf/spacy to generalise the title extraction since there is same context. But I see it is working only when I ask queries with same title I used in training like 

> I would like to create a task with title test title
> Create a task with title test title

If I ask like following then title is not getting extracted

> I would like to create a task with title issue resolution
> Create a task with title operations review

I can not train with all titles since user can input any text. The same scenario applies to description and many other such attributes.

How to handle this scenario ? 

@tmbo @amn41 
Can I get some help on this ?

**Comment by anshulsolanki** [comment]
your crf model is getting overfit because it is never seeing title other than 'test'.
I would recommend you training with same samples but replacing 'test' with some more variations.

Moreover if 'XXX title' is the way user will always write title, you can extract the same using regular expression.

**Comment by madhusudhannsn** [comment]
@anshulsolanki Thanks for the reply. I tried training with different titles but I see no change in the accuracy. I can not use regular expressions as the format may not be 'XXX title'.

**Comment by anshulsolanki** [comment]
If possible please share your training file. Some suggestion might be given looking at data and how is it tagged.

**Comment by madhusudhannsn** [comment]
@anshulsolanki Sorry for the late reply. Please find some of training examples

```
 Create a task with title Team outing at Phoenix Mall
 Could you please delete the task titled sample integration
 Change the due date of a task titled sample integration test
 Create a task with title Daily standup meeting
 Create a task with title Daily phone call
 Create a task with title Holiday
 Create a task with title Happy Birthday
 Create a task with title Team outing
 Create a task with title daily task
 Create a task with title good times
 Create a task with title my life
 Create a task with title my vacation
 Create a task with title sample integration
 I would like to change the due date of task titled sample integration
 I would like to change the due date of task with title sample integration
 I would like to delete task with title sample integration
```

This is not like a normal scenario where entity extractor predicts something similar to the entity values in the training examples. The issue here is extractors will see complete new values at prediction and they have to generalise prediction based on context. I can use regular expressions but user may or may not follow my patterns. 

@wrathagom @tmbo @amn41 

We have tried optimised ner_crf and spacy model trained on our data. But we see only 20% accuracy.
https://sklearn-crfsuite.readthedocs.io/en/latest/tutorial.html#hyperparameter-optimization

Could you guys please help me on this issue.

**Comment by anshulsolanki** [comment]
you should not be using ner for this.
you should use dependency parsing.

I was able to extract title correctly for each of your examples using dependency parsing.
Have a look at https://demos.explosion.ai/displacy/

**Comment by madhusudhannsn** [comment]
@anshulsolanki Thanks for the suggestion and it works for 50% of my examples. It solves cases where title contain only nouns but fails if title has any verbs, prepositions, conjunctions etc., For example

```
Create a meeting with title Build broken on current develop
I would like to change the due date of task with title Staging updated
I would like to change the due date of task with title call with Nathan
Create a task with title my vacation
Create a task with title Team outing at Phoenix Mall
```

**Comment by anshulsolanki** [comment]
@madhusudhannsn , good to know that you have moved from 20% to 50%.
For the examples you have posted, if you clearly look at the parse tree structure, it is evident that the structure is very well defined and you can reach the entity title by walking the tree.

E.g. for "Create a task with title my vacation"
"my vacation" is a direct object for root verb "create". Though "a task" is also a direct object, but a simple Drool based rule engine can filter that out.
Moreover you can use a combination of dependency parsing + NER + rule engine to reach your goal.

or use dependency parsing , then feed the auto annotated text to NER for train ( and eventually predict).

No single silver bullet will solve this.


Let me know how it goes. I can help to fine tune your approach if needed

**Comment by madhusudhannsn** [comment]
@anshulsolanki  I think with the above combination, I can achieve good accuracy. Thank you so much for the help.

**Comment by sameerkumar16** [comment]
@anshulsolanki  how to incorporate dependency parsing in Rasa_nlu .I mean it's not in the pipeline and how will the entity be parsed .

**Comment by anshulsolanki** [comment]
@sameerkumar16  , How to add dependency parsing to pipeline is something that will have to be explored. I have not done it myself, having said that it should not be difficult. Spacy is already available in pipeline and you can use the dependency parser provided by Spacy.

If i do it , i will let you.
